src/EnergyPlus/NodeInputManager.hh

```diff
@@ -76,6 +76,7 @@ namespace NodeInputManager {
         Primary = 1,
```

Comment:
Do these have to start at 1? Why can they not start at 0?

Reply:
Here is my tentative answer on this: 
According to @mjwitte 's explanation in a comment above, this numbering actually had occurred long time ago and had been the mere identification for quite a while; and the `enum` was a more recent addition to map the actual numbers. I would probably interpret the `enum` names (Primary to Quaternary) as the "keys" the existing numbers instead, with the current setting of:
```
    enum class CompFluidStream
    {
        Invalid = -1,
        Primary = 1,
        Secondary = 2,
        Tertiary = 3,
        Quaternary = 4,
        Num
    };
```

Actually it seems the actual numbering is more important the `enum` name. One problem I see when changing the numbering (to start from `0`) is that there are many code lines are actually dependent on actual numbers and then there could probably be a lot more code lines that need to be examined and changed.

Reply:
By the last statement you mean that the code is written like `== 1` as opposed to `== CompFluidStream::Primary`? Yikes!

Reply:
I believe so---the older code probably originally just used `==1`. Actually there are still some mixed uses in the current code: some are like `CompFluidStream::Primary`---the new way; but there are quite some places where the numbers are directly used (with a static cast to `enum` though), such as `static_cast<NodeInputManager::CompFluidStream>(ConnectionLoop + 2)` shown in the code block mentioned in [this comment](https://github.com/NREL/EnergyPlus/pull/9364#discussion_r844074385) above.

Reply:
As the original author of the node connections, this was intended to be an integer tag, simply to group nodes that belong to the same fluid stream within a given component. It's used for connection checking and it's also written as integers to the bnd output for HVAC Diagram. There's really no limit the number of possible fluid streams, so converting this to a fixed set of enum values made the calls to `getOnlySingleNode` look nice, but it broke this place where it's based on looping through a set of nodes and incrementing the fluidstream.

It's stored as an enum in `state.dataBranchNodeConnections->NodeConnections(Count).FluidStream` so I think any values (including zero) would be safe to use as far as EnergyPlus is concerned. But it could trip up places where it's interpreted from a loop index as @jcyuan2020 noted. It's bit tricky to scan the code for cases where this argument is set through a `cast`.

And I'm not sure if HVACDiagram would choke on zero values or not.

In looking at the code, there's an optional `IncrementFluidStream` in `NodeInputManager::GetNodeNums` that will fail if it's called with a node list of more than 3 (or now 4) nodes. See [code here](https://github.com/NREL/EnergyPlus/blob/develop/src/EnergyPlus/NodeInputManager.cc#L194-L196). Based on a quick check, it looks like [outside air nodes](https://github.com/NREL/EnergyPlus/blob/c2a7f6078341696d5b1d26736216175c622072be/src/EnergyPlus/OutAirNodeManager.cc#L216-L234) are the only place that use the `IncrementFluidStream` argument. I'm surprised this doesn't fail the integration tests, but maybe we don't have any inputs that use a nodelist name in that object?  `IncrementFluidStream` is also used [here](https://github.com/NREL/EnergyPlus/blob/c2a7f6078341696d5b1d26736216175c622072be/src/EnergyPlus/OutAirNodeManager.cc#L278-L290).

Reply:
@amirroth I guess I don't fully (partly?) understand what happens with enums at run-time. Based on the above comments, this input from _SmallOffice_Dulles.idf violates the enum list, because it will set the fluid stream values to 1,2,3,4:
```
  NodeList,
    OutsideAirInletNodes,    !- Name
    OAInletNode_SouthZone_1stFloor,  !- Node 1 Name
    OAInletNode_NorthZone_1stFloor,  !- Node 2 Name
    OAInletNode_SouthZone_2ndFloor,  !- Node 3 Name
    OAInletNode_NorthZone_2ndFloor;  !- Node 4 Name
```
Are enums only validated at compile time?

Reply:
- To @mjwitte's question first. I assume that when you say "validation" you mean "type-checking".  In C++, type-checking happens at compile-time. In C/C++, `enum` is a subtype of `int`. As with any sub-typing relationship, you can safely use an `enum` as an `int` (i.e., use an `enum` anywhere that expects an `int` like an index into an array) and you can even do so without casting but not the other way around (i.e., use an `int` in a place that expects an `enum`). In C++, an `enum class` is not technically a subtype of `int` but it is for all intents and purposes and you can safely use an `enum class` as an `int` although you have to add `static_cast<int>(compFluidStream)`. However, you still cannot use an `int` as an `enum class` and so `static_cast<enum class CompFluidStream>(i)` is not a safe thing to do. By using `static_cast` you are telling the compiler to let you do this unsafe thing anyway because "you know what you're doing" even though in this case you obviously don't. In this specific case, you cannot take arbitrary integers and safely use them as `enum class CompFluidStream`. If something doesn't break, it's just luck.
- What to do in this case? If `CompFluidStream` can really be an arbitrary integer then it should be an `int` and not an `enum class`. If you want to use `enum class` for some reason, then you should check in the `NodeList getInput` function that the range of this `enum class` is not violated.